### PR TITLE
HPCC-24155 Update Kafka plugin to support UTF-8 encoded messages

### PR DESCRIPTION
**Pull request by dcamper**
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Added UTF8-specific regression tests.  All existing and new tests pass on all engines.

### CODE REVIEW

File: plugins/kafka/kafka.cpp [hunk 1]
```diff
@@ -1031,9 +1033,10 @@ namespace KafkaPlugin
 // Plugin Initialization and Teardown
 //==============================================================================
 
-#define CURRENT_KAFKA_VERSION "kafka plugin 1.0.0"
+#define CURRENT_KAFKA_VERSION "kafka plugin 1.1.0"
 
 static const char* kafkaCompatibleVersions[] = {
+    "kafka plugin 1.0.0",
```
Comment (dcamper) [inline plugins/kafka/kafka.cpp:1039]: I am unsure if changes made in this PR resulted in an incompatible module or not.  If it is incompatible, this line should be removed.  Feedback welcome.

**Comment by dcamper** [comment]
@richardkchapman Please review or assign for review.  Thanks!

File: plugins/kafka/kafka.cpp (hunk 1 above)
Comment (richardkchapman) [reply plugins/kafka/kafka.cpp:1039]: To be honest the compatible versions mechanism is pretty redundant these days, as we ship plugins in lock-step with platform versions and expect them to match.

**Comment by richardkchapman** [comment]
I think in the interests of existing ECL code continuing to compile, I would expect to add new functions taking/returning UTF8 rather than replacing the exiting ones.

@ghalliday Any comments?

**Comment by dcamper** [comment]
New compiles should Just Work, as evidenced by the original (unchanged) regression tests.  I am curious how the change would affect something like deployed ROXIE queries, though.

**Comment by richardkchapman** [comment]
They would fail to load due to unresolved symbol, if they called publishMessage with the old parameter mangling.

You can avoid this by leaving the old function in place (calling on to the new, if that's simplest)

Something like:

ECL_KAFKA_API bool ECL_KAFKA_CALL publishMessage(const char* brokers, const char* topic, const char* message, const char* key)
{
return publishMessage(nullptr, brokers, topic, strlen(message), message, strlen(key, key);
}

You'd need to ensure that the function you call on to would accept ctx being null .

Incidentally, should brokers and topic also be utf8 ?

**Comment by dcamper** [comment]
Excellent point, Richard.  I should have done that to begin with.  I will make that change now.

**Comment by dcamper** [comment]
FYI, the previous code version actually had an incorrect header method signature for publishMessage().  The .ecllib file cites a context, the header did not, but the actual method did include a context parameter.  So the previous code advertised a method that did not actually exist while including a concrete method that just happened to match what callers were expecting.  TL;DR, a null contextPtr isn't a problem.

**Comment by richardkchapman** [comment]
A null context will crash:

        std::call_once(pubCacheInitFlag, setupPublisherCache, ctx->queryContextLogger().queryTraceLevel());

**Comment by dcamper** [comment]
Yes, but it doesn't happen; that's what I'm saying.  The advertised signature in the header (the one without ctx) is never referenced.

**Pull request merged**